runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/StreamingEngineClient.java

```diff
@@ -0,0 +1,432 @@
+    // We only want to set and start this value once.
+    if (getWorkerMetadataStream == null) {
+      synchronized (this) {
+        if (getWorkerMetadataStream == null) {
+          getWorkerMetadataStream =
+              streamFactory.createGetWorkerMetadataStream(
+                  dispatcherClient.getDispatcherStub(),
+                  getWorkerMetadataThrottleTimer,
+                  endpoints ->
+                      // Run this on a separate thread than the grpc stream thread.
+                      consumeWorkerMetadataExecutor.submit(
+                          () -> consumeWindmillWorkerEndpoints(endpoints)));
+        }
+      }
+    }
+  }
+
+  private synchronized ImmutableMap<Endpoint, WindmillConnection> createNewWindmillConnections(
+      ImmutableSet<Endpoint> newWindmillEndpoints) {
+    ImmutableMap<Endpoint, WindmillConnection> currentWindmillConnections =
+        connections.get().windmillConnections();
+    Map<Endpoint, WindmillConnection> newWindmillConnections =
+        currentWindmillConnections.entrySet().stream()
+            .filter(entry -> newWindmillEndpoints.contains(entry.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // Reuse existing stubs if they exist.
+    newWindmillEndpoints.forEach(
+        newEndpoint ->
+            newWindmillConnections.putIfAbsent(
+                newEndpoint, WindmillConnection.from(newEndpoint, this::createWindmillStub)));
+
+    return ImmutableMap.copyOf(newWindmillConnections);
+  }
+
+  private synchronized ImmutableMap<WindmillConnection, WindmillStreamSender>
+      closeStaleStreamsAndCreateNewStreams(
+          ImmutableSet<WindmillConnection> newWindmillConnections) {
+    ImmutableMap<WindmillConnection, WindmillStreamSender> currentStreams =
+        connections.get().windmillStreams();
+
+    // Close the streams that are no longer valid.
+    currentStreams.entrySet().stream()
+        .filter(
+            connectionAndStream -> !newWindmillConnections.contains(connectionAndStream.getKey()))
+        .map(Map.Entry::getValue)
+        .forEach(WindmillStreamSender::closeAllStreams);
+
+    return newWindmillConnections.stream()
+        .collect(
+            toImmutableMap(
+                Function.identity(),
+                newConnection ->
+                    Optional.ofNullable(currentStreams.get(newConnection))
+                        .orElseGet(() -> createWindmillStreamSenderWithNoBudget(newConnection))));
+  }
+
+  private ImmutableMap<Endpoint, Supplier<GetDataStream>> createNewGlobalDataStreams(
+      ImmutableSet<Endpoint> newGlobalDataEndpoints) {
+    ImmutableMap<Endpoint, Supplier<GetDataStream>> currentGlobalDataStreams =
+        connections.get().globalDataStreams();
+
+    return newGlobalDataEndpoints.stream()
+        .map(endpoint -> existingOrNewGetDataStreamFor(endpoint, currentGlobalDataStreams))
+        .collect(toImmutableMap(Pair::getKey, Pair::getValue));
```

Comment:
instead of pairs how about using Function.identity() like above?

newGlobalDataEndpoints.stream().collect(
  toImmutableMap(Function.identity(), existingOrNewGetDataStreamFor(endpoint, currentGlobalDataStreams));

Reply:
done! good find